Editorial for 412D (#6342)
cpinitiative/usaco-guide: solutions/silver/cf-412D.mdx

---
id: cf-412D
source: CF
title: Giving Awards
author: John Zhou
---

[Official Editorial](https://codeforces.com/blog/entry/11756)

## Explanation

Since there are no restrictions of both $u,v$ and $v,u$, it is always possible for $a_i$ to follow $b_i$ either directly or indirectly in each restriction; if we make this happen all the time (except in cyclic situations as described below), then we’d get a working configuration.

> Construct a directed graph with edges $a_i\to b_i$, where we call $b_i$ the neighbor of $a_i$ for each $i$.

Author: ```suggestion
Construct a directed graph with edges $a_i\to b_i$, where we call $b_i$ the neighbor of $a_i$.
```

> It is possible to make all nodes come after all its neighbors if there is no cycles in this graph.

Author: ```suggestion
It is possible to make each node follow all its neighbors if there are no cycles.
This will lead to a valid configuration, as an element cannot follow another directly if it is before the latter element in the first place.
```

> If there is, we can neglect an edge in a cycle, which still gives a viable configuration, as the cycles are at least of size $3$, meaning the two vertices associated with the edge we break will not be adjacent in the first place, so their direction does not matter.

Reviewer: i get what you mean but imo the wording in this paragraph is rlly dense and not very easy to digest - also saying breaking an edge implies we'd modify the graph but i think you meant ignore

> If there is, we can neglect an edge in a cycle, which still gives a viable configuration, as the cycles are at least of size $3$, meaning the two vertices associated with the edge we break will not be adjacent in the first place, so their direction does not matter.

Author: ```suggestion
If there is a cycle, we can neglect an edge in a cycle. This is still valid, as the cycles are at least $3$ vertices.

To see why, we use an example: we require vertex A to placed before B, B to be before C, and C to be before A; from the first two requirements, A must be before C with at least $1$ element (B) in between. Since A and C are not adjacent, their order does not matter, so we may neglect the third edge.
```

> We use a flood fill DFS to traverse all the nodes; at each node, we DFS all its neighbors before outputting the current node number.

Reviewer: pretty sure its postorder dfs not flood fill dfs

> We use a flood fill DFS to traverse all the nodes; at each node, we DFS all its neighbors before outputting the current node number.

Author: ```suggestion
We use a post order DFS to traverse all the nodes; at each node, we DFS all its neighbors before outputting the current node number.
```

This ensures that all neighbors of a vertex come before itself.

> In case of a cycle, we would revisit an already-visited vertex, which we can safely skip as breaking an edge is permissible.

Reviewer: same wording problem here - "breaking" shd be ignoring

> In case of a cycle, we would revisit an already-visited vertex, which we can safely skip as breaking an edge is permissible.

Author: ```suggestion
In case of a cycle, we would revisit an already-visited vertex, which we can safely skip as ignoring an edge is permissible.
```

---
id: cf-412D
source: CF
title: Giving Awards
author: John Zhou
---

[Official Editorial](https://codeforces.com/blog/entry/11756)

## Explanation

> Since there are no restrictions of both $u,v$ and $v,u$, it is always possible for $a_i$ to follow $b_i$ either directly or indirectly in each restriction; if we make this happen all the time (except in cyclic situations as described below), then we’d get a working configuration.

Reviewer 2: what do you mean follow indirectly? like just after it in the array?

Author: Nah first clause is also kinda redundant since that case is covered later.
```suggestion
Having each $a_i$ goes after its $b_i$ in the list results in a working configuration, as $b_i$ cannot directly follow $a_i$ when it is before $a_i$ in the first place.
This can always happen, except in situations where the graph formed by connecting each $a_i$ to $b_i$ has cycles.
We call each $b_i$ the neighbor of $a_i$.
```

Construct a directed graph with edges $a_i\to b_i$, where we call $b_i$ the neighbor of $a_i$.
It is possible to make each node follow all its neighbors if there are no cycles.
This will lead to a valid configuration, as an element cannot follow another directly if it is before the latter element in the first place.
If there is a cycle, we can neglect an edge in a cycle. This is still valid, as the cycles are at least $3$ vertices.

> To see why, we use an example: we require vertex A to placed before B, B to be before C, and C to be before A; from the first two requirements, A must be before C with at least $1$ element (B) in between. Since A and C are not adjacent, their order does not matter, so we may neglect the third edge.

Reviewer 3: this explanation doesn't convince me that the solution is correct. what if there are more cycles in the graph, which then require A->B or B->C to be removed?

Author: orz orz orz

This would not happen in an actual DFS—we DFS A->B and B->C first before breaking C->A and by that point AB and BC is already guaranteed.  A different cycle back to A will break a different edge when A is marked as visited as AB is already traversed

what do you think?? Is the explanation above reasonable logically?  If so I’ll polish it and add it in an Info component as this soln is already mostly guessforces anyways

Reviewer 3: idk, it's still not really convincing that this argument covers all possible cases

imo the most convincing way to show correctness is to rely on properties of the DFS tree introduced in this (not Silver) module: https://usaco.guide/adv/BCC-2CC?lang=cpp#resource-https%3A%2F%2Fcodeforces.com%2Fblog%2Fentry%2F68138

Author: IMHO intuition behind this trick is a tad more useful than the formal proof

would you be against quoting the reference and explaining how to use DFS tree to prove in a spoiler?

Reviewer 3: That should be okay. Please include at top of the internal sol a note about the official editorial solution being very different, and that your solution is not necessarily in scope for USACO Silver.

Author: Thanks.  Tag me tomorrow if I don’t do this by then.

Also another TODO is to open an issue with writing the simplified solution (which I think is way too much guessforces)

Reviewer 2: I'll actually add solution 1 cuz I dont really like solution 2

Author: Looks like you've already done so -- I've been planning to do these changes today.  In that case, tysm!

</Warning>

## Explanation

We consturct the directed graph with edges $a_i\to b_i$, and call each $b_i$ the neighbor of $a_i$.
Having each $a_i$ goes after its $b_i$ in the list results in a working configuration, as $b_i$ cannot directly follow $a_i$ when it is before $a_i$ in the first place.
This can always happen, except when the graph has cycles.
If there is a cycle in the graph, however, we can neglect an edge in a cycle. This is still valid, as the cycles have at least $3$ vertices.

To see why, consider this example: we require vertex A to placed before B, B to be before C, and C to be before A; from the first two requirements, A must be before C with at least $1$ element (B) in between. Since A and C are not adjacent, their order does not matter, so we may neglect the third edge.

We use a post order DFS to traverse all the nodes; at each node, we DFS all its neighbors before outputting the current node number.
This ensures that all neighbors of a vertex come before itself.
In case of a encountering a cycle while traversing, we would revisit an already-visited vertex, which we can safely skip as neglecting an edge is permissible.

> Note that by DFS'ing each unvisited vertex in order, we may start visiting a neighbor of some other vertex we have not yet traversed; however, that other vertex would be traversed in a later DFS, so the order is still correct.  The later DFS will also skip the current neighbor, as it has already been traversed.

Reviewer: ```suggestion
Note that by DFS'ing each unvisited vertex in order, we may visit a neighbor of some other vertex we have not yet visited. However, this is not an issue, as that other vertex would be traversed in a later DFS, so the order is still correct. Its DFS will simply skip the already visited neighbor
```

## Implementation

**Time complexity:** $\mathcal O(n+m)$

<LanguageSection>

<CPPSection>

```cpp

> // Based on solution by JuanMata on CodeForces

Reviewer 4: is this necessary?

Author: Let’s give credit cause I copied soln off of codeforces

Author: (I copied the code and changed it

Reviewer 4: > Let’s give credit cause I copied soln off of codeforces

What about editorial then? Do you wrote it yourself or used LLM?

Author: I wrote editorial text myself but used the implementation from a codeforces submission

here
```suggestion
// Based on code by JuanMata on CodeForces
```

---
id: cf-412D
source: CF
title: Giving Awards
author: John Zhou
---

# Solution 1

[Official Editorial](https://codeforces.com/blog/entry/11756)

# Solution 2

<Warning>

The correctness of solution 2 is mostly easily proven using information about DFS trees from [this resource](/adv/BCC-2CC?lang=cpp#resource-https%3A%2F%2Fcodeforces.com%2Fblog%2Fentry%2F68138). However, this is outside the scope of USACO Silver.

</Warning>

## Explanation

> Since there are no restrictions of both $u,v$ and $v,u$, it is always possible for $a_i$ to follow $b_i$ either directly or indirectly in each restriction; if we make this happen all the time (except in cyclic situations as described below), then we’d get a working configuration.

Reviewer 2: are you just trying to say it is always possible for a_i to be after b_i or trying to say that a_i and b_i can always touch?

Also, you have used u, v in the first part of the sentence and then transition to a_i, b_i  when it seems like both of these are just refering to a restriction, so I would try to keep it consistent.

Author: The clause is cut and I clarified wording

> Construct a directed graph with edges $a_i\to b_i$, where we call $b_i$ the neighbor of $a_i$.

Reviewer 2: ```suggestion
Construct a directed graph with edges $a_i\to b_i$, and call $b_i$ the neighbor of $a_i$.
```

> It is possible to make each node follow all its neighbors if there are no cycles.

Reviewer 2: so just double checking, but here by "follow" you meant reside after all it's neighbors in the array?

Author: Yep.  But I'm gonna cut this paragraph

> This will lead to a valid configuration, as an element cannot follow another directly if it is before the latter element in the first place.

Reviewer 2: Since this kinda feels a bit like weird sounding (lines 25-26) one way I would propose to make it clear is

If there are no cycles, by making each node follow all of it's neighbors, we are left with a valid configuration. For two elements in a cycle, ..... (something like it can't be after some neighbor will being required to be before it due to the cycle meaning, so a cycle is not possible ...).

My feedback is: (feel free to look at it however you want)
1. The transition from line 25-26 seems a bit weird to me like you are saying this will lead to a configuration implying line 25 was some sort of if we do ..., but line 25 was an assertion saying we have a valid configuration if this is true.
 
2. The part of line 26 when you explain "if it is before the latter element in the first place" seems a little unclear, because you don't really reference the cycle is why it is technically supposed to be before the latter element in the first place.

Really sorry for being hypercritical 😭

Author: Line 26 is about the no-cycles case.  I'm gonna cut this whole paragraph and tighten

If there is a cycle, we can neglect an edge in a cycle. This is still valid, as the cycles are at least $3$ vertices.

To see why, we use an example: we require vertex A to placed before B, B to be before C, and C to be before A; from the first two requirements, A must be before C with at least $1$ element (B) in between. Since A and C are not adjacent, their order does not matter, so we may neglect the third edge.

We use a post order DFS to traverse all the nodes; at each node, we DFS all its neighbors before outputting the current node number.
This ensures that all neighbors of a vertex come before itself.

> In case of a cycle, we would revisit an already-visited vertex, which we can safely skip as ignoring an edge is permissible.

Reviewer 2: In the case of encountering a cycle while traversing, we would revisit an already-visited vertex, which we can safely skip as ignoring an edge is allowable.
Is this what you meant? Also to clarify for myself, is ignoring an edge permissible according to that logic you had above?

Author: Yes and yes
```suggestion
In case of a encountering a cycle while traversing, we would revisit an already-visited vertex, which we can safely skip as neglecting an edge is permissible.
```

> We start DFS’ing from each node of it is not already visited.

Reviewer 3: ^ don't understand what this is saying

same for the next sentence as well

Reviewer 2: Likewise for 35, 36-37 can prolly be simplified cuz I kinda understand what you're saying??

Author: Simplified a lot here.

> Note that it is possible for a non-visited node to be a neighbor of some other vertex; however, we still DFS it and mark it as visited, as the other vertex would be visited later in another DFS, where the current vertex will be skipped.

Author: ```suggestion
Note that we may start visiting a neighbor of some other vertex we have not yet traversed; however, that other vertex would be traversed in a later DFS, so the order is still correct.  The later DFS will also skip the current neighbor, as it has already been traversed.
```

> That is still correct, as we’ve already done the DFS on that vertex now.

Author: ```suggestion
```

## Implementation

**Time complexity:** $\mathcal O(n+m)$

<LanguageSection>

<CPPSection>

```cpp
// Based on solution by JuanMata on CodeForces
#include <cstring>
#include <iostream>
#include <vector>

using namespace std;

vector<int> graph[30005];
bool visited[30005];

void dfs(int node) {
	// skip if visited
	if (visited[node]) { return; }
	visited[node] = true;

	// visit neighbors
	for (auto next : graph[node]) { dfs(next); }

	cout << node << ' ';
}

int main() {
	int n, m;
	cin >> n >> m;

	for (int i = 0; i < m; i++) {
		int u, v;
		cin >> u >> v;
		graph[u].push_back(v);
	}

> 	memset(visited, 0, sizeof(visited));

Reviewer 2: Is there a reason for memset? I mean i dont care although I was js curious.

Author: Being chinese they teach you this method... they also encourage memset(0x7f) to do infinity.

Author: but sincey ou brought it up imma edit it

Author: Hold your horses the real reason is because I copied this code from a public CodeForces submission someone shared and then heavily edited it; this was leftovers.

	for (int i = 1; i <= n; i++) {
		if (!visited[i]) { dfs(i); }
	}

	cout << endl;

> 	return 0;

Reviewer 2: Also yeah Idk if you need this

Author: ```suggestion
```

this was for nostalgia
